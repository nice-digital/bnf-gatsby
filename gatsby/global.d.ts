// Let TypeScript know we're importing SCSS (and SCSS modules)
// Avoids "Can't import CSS/SCSS modules. TypeScript says “Cannot Find Module”"
// See https://stackoverflow.com/a/41946697/486434
declare module "*.module.scss" {
	const content: { [className: string]: string };
	export default content;
}

// So that we can import JSON files in TypeScript.
declare module "*.json" {
	const value: unknown;

	export default value;
}
