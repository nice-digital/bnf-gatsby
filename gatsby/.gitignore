node_modules/
.cache/
public

# dotenv environment variables file
.env
.env.*
