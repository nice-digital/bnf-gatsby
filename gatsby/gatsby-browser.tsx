import {
	ShouldUpdateScrollArgs,
	type RouteUpdateArgs,
	type WrapPageElementBrowserArgs,
} from "gatsby";
import { type ReactElement } from "react";

import { Layout } from "@/components/Layout/Layout";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
	// Default return statement when prevLocation is falsy
	return;
};

export const wrapPageElement = ({
	element,
	props,
}: WrapPageElementBrowserArgs): ReactElement => (
	<Layout {...props}>{element}</Layout>
);

/**
 * Gatsby hook for overriding scroll position
 * See https://www.gatsbyjs.org/docs/browser-apis/#shouldUpdateScroll
 */
export const shouldUpdateScroll = ({
	routerProps: { location },
	prevRouterProps,
	getSavedScrollPosition,
}: ShouldUpdateScrollArgs): boolean | string | [number, number] => {
	if (
		// If there's no previous route props we're coming from an external site, which means
		// we want to scroll to the hash (if there is one), and _not_ a saved scroll position
		!prevRouterProps ||
		// Or we're linking to as hash within the same page
		(prevRouterProps.location.pathname === location.pathname && location.hash)
	) {
		// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = location.hash
					? document.querySelector(location.hash)
					: null;

				if (targetElement) {
					targetElement.setAttribute("tabIndex", "-1");
					(targetElement as HTMLElement).focus();
					targetElement.scrollIntoView();
				}
			});
		});
		return false;
	}

	const savedScrollY = (getSavedScrollPosition(location)?.[1] || 0) as number;
	if (savedScrollY > 0) {
		window.scrollTo(0, savedScrollY);
		return false;
	}

	if (location.hash) {
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = document.querySelector(location.hash);

				// Default to Gatsby's default behaviour if the element doesn't exist
				if (!targetElement) return true;

				// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
				targetElement.setAttribute("tabIndex", "-1");
				(targetElement as HTMLElement).focus();
				targetElement.scrollIntoView();
				return false;
			});
		});
		return false;
	}

	requestAnimationFrame(() => {
		requestAnimationFrame(() => {
			const contentStartElement = document.getElementById("content-start");

			if (!contentStartElement) return true;

			contentStartElement.setAttribute("tabIndex", "-1");

			//scrolling to 0,0 to preserve existing BNF navigation behaviour
			window.scrollTo(0, 0);
			return false;
		});
	});
	// Default to scrolling to the content start element as the standard navigation behaviour

	return true;
};
