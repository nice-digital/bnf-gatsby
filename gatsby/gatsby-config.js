require("dotenv").config({
	path: `.env.${process.env.NODE_ENV}`,
});

const isBNF = process.env.GATSBY_SITE === "bnf";

if (!process.env.GATSBY_SEARCH_URL)
	throw new Error(
		"Env var GATSBY_SEARCH_URL isn't set. Did you forget to add it?"
	);

module.exports = {
	jsxRuntime: "automatic",
	siteMetadata: {
		isBNF,
		siteUrl: `https://${isBNF ? "bnf" : "bnfc"}.nice.org.uk`,
		siteTitleShort: isBNF ? "BNF" : "BNFC",
		siteTitleLong:
			"British National Formulary" + (isBNF ? "" : " for Children"),
	},
	plugins: [
		{
			resolve: `gatsby-source-bnf`,
			options: {
				feedURL: process.env.FEED_URL,
				userKey: process.env.FEED_USER_KEY,
				site: isBNF ? "bnf" : "bnfc",
			},
		},
		// Avoid errors like "ModuleNotFoundError: Module not found: Error: Can't resolve '@/components/Layout/Layout'" when using custom paths in tsconfig.json
		`gatsby-plugin-tsconfig-paths`,
		{
			resolve: `gatsby-plugin-sass`,
			options: {
				// Make CSS modules play nicely with TypeScript, see also global.d.ts for *.scss TS def
				cssLoaderOptions: {
					esModule: false,
					modules: {
						namedExport: false,
					},
				},
			},
		},
		"gatsby-plugin-catch-links",
		"gatsby-plugin-react-helmet",
		"gatsby-plugin-remove-generator",
		{
			resolve: `gatsby-plugin-sitemap`,
			options: {
				excludes: [`/search/`],
			},
		},
		{
			resolve: "gatsby-plugin-google-tagmanager",
			options: {
				id: "GTM-5H5L9GK",
				includeInDevelopment: true,
			},
		},
		{
			resolve: "gatsby-plugin-manifest",
			options: {
				icon: "src/images/icon.png",
			},
		},
	],
};
