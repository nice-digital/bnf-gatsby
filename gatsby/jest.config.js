const { pathsToModuleNameMapper } = require("ts-jest");

const baseConfig = require("./../jest.config"),
	{ compilerOptions } = require("./tsconfig.json");

// Translation of TypeScript path mappings to jest paths: https://kulshekhar.github.io/ts-jest/docs/getting-started/paths-mapping
const tsPathsModuleNameMappings = pathsToModuleNameMapper(
	compilerOptions.paths,
	{ prefix: "<rootDir>/../" }
);

module.exports = {
	// Extend the base config from the root as it's got shared/common config
	...baseConfig,
	displayName: "gatsby",
	rootDir: "./",
	roots: ["<rootDir>"],
	projects: undefined,
	moduleNameMapper: {
		"\\.(css|scss|svg)$": "identity-obj-proxy",
		// Alias React/Next into the gatsby folder because of our monorepo setup.
		// Avoids errors like "multiple instances of React" with hooks
		"^react(.*)$": "<rootDir>/node_modules/react$1",
		"^gatsby(.*)$": "<rootDir>/node_modules/gatsby$1",
		// Aliases to match paths in tsconfig.json
		...tsPathsModuleNameMappings,
	},
	transformIgnorePatterns: ["/node_modules/", "/dist/"],
	setupFilesAfterEnv: ["./src/jest.setup.ts"],
	testPathIgnorePatterns: ["./config/"],
	testURL: "https://bnf-gatsby-tests.nice.org.uk",
	testEnvironment: "jsdom",
	globals: {
		"ts-jest": {
			tsconfig: {
				...compilerOptions,
				// To avoid errors like "Class constructor App cannot be invoked without 'new'"
				target: "ESNext",
				jsx: "react-jsxdev",
			},
		},
	},
};
