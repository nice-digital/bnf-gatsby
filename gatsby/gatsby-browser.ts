import { RouteUpdateArgs } from "gatsby";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
};
