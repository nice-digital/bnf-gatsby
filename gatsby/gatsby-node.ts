import { CreateSchemaCustomizationArgs, CreateWebpackConfigArgs } from "gatsby";

import { initialFieldExtension } from "./src/field-extensions/initial";

/**
 * Gatsby hook for customizing the schema.
 * See https://www.gatsbyjs.org/docs/schema-customization/
 */
export const createSchemaCustomization = ({
	actions: { createFieldExtension, createTypes },
}: CreateSchemaCustomizationArgs): void => {
	createFieldExtension(initialFieldExtension);

	// Add initials for types used in A-Z lists, as it's easier for grouping
	// We don't add the initial field in the source plugin because grouping is a presentation concern
	const typeDefs = `
	  type BnfDrug implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfTreatmentSummary implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfInteractant implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	`;
	createTypes(typeDefs);
};

/**
 * Gatsby hook for overriding webpack config
 * See https://www.gatsbyjs.com/docs/how-to/custom-configuration/add-custom-webpack-config/
 */
export const onCreateWebpackConfig = ({
	stage,
	actions,
	getConfig,
	plugins,
}: CreateWebpackConfigArgs): void => {
	// Disable order warnings from mini CSS plugin
	// See https://github.com/gatsbyjs/gatsby/discussions/30169#discussioncomment-621285
	if (stage === "build-javascript" || stage === "develop") {
		const config = getConfig();

		const miniCss = config.plugins.find(
			(plugin: () => void) => plugin.constructor.name === "MiniCssExtractPlugin"
		);

		if (miniCss) {
			miniCss.options.ignoreOrder = true;
		}

		actions.replaceWebpackConfig(config);

		actions.setWebpackConfig({
			plugins: [plugins.provide({ process: "process/browser" })],
		});
	}
};
