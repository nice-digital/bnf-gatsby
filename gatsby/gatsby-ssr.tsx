import { PreRenderHTMLArgs } from "gatsby";
import React from "react";

export const onPreRenderHTML = ({
	getHeadComponents,
	replaceHeadComponents,
}: PreRenderHTMLArgs): void => {
	const components = getHeadComponents();

	components.push(
		<script
			src={process.env.GATSBY_COOKIE_BANNER_URL}
			type="text/javascript"
			key="cookie-banner"
			async
		></script>
	);

	replaceHeadComponents(components);
};
