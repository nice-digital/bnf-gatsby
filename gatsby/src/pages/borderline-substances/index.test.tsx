import { render, screen, waitFor } from "@testing-library/react";
import { useStaticQuery } from "gatsby";

import { mockBorderlineSubstancesPagesQueryData } from "./{BnfBorderlineSubstancesTaxonomy.slug}.test";

import BorderlineSubstanceIntroductionPage, {
	type BorderlineSubstanceIntroductionPageProps,
} from "./";

const props: BorderlineSubstanceIntroductionPageProps = {
	data: {
		bnfBorderlineSubstancesIntroduction: {
			title: "Borderline Substances",
			sections: [],
		},
	},
};

describe("BorderlineSubstanceIntroductionPage", () => {
	beforeEach(() => {
		(useStaticQuery as jest.Mock).mockReturnValue(
			mockBorderlineSubstancesPagesQueryData
		);
	});

	it("should match snapshot for page contents", () => {
		const { container } = render(
			<BorderlineSubstanceIntroductionPage {...props} />
		);
		expect(container).toMatchSnapshot();
	});

	it("should set page title", async () => {
		render(<BorderlineSubstanceIntroductionPage {...props} />);

		await waitFor(() => {
			expect(document.title).toStartWith("Borderline Substances | ");
		});
	});
});
