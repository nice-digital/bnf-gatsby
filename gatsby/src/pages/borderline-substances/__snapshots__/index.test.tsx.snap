// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BorderlineSubstanceIntroductionPage should match snapshot for page contents 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="https://www.nice.org.uk/"
        >
          NICE
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/"
        >
          BNF
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Borderline Substances
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "name": "Borderline Substances"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
    id="content-start"
  >
    <h1
      class="page-header__heading"
    >
      <span>
        Borderline Substances
      </span>
    </h1>
  </div>
  <div
    class="grid grid--loose"
    data-testid="body"
  >
    <div
      class="hide-print"
      data-g="12 md:4 lg:3"
    >
      <nav
        aria-label="Borderline substances"
        class="stacked-nav"
      >
        <p
          class="stacked-nav__root"
        >
          <a
            aria-current="false"
            href="/borderline-substances/"
          >
            Borderline substances
          </a>
        </p>
        <ul
          class="stacked-nav__list"
        >
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/parent-1/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Parent taxonomy 1
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/taxonomy-1/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Sub taxonomy 1
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/taxonomy-2/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Sub taxonomy 2
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/parent-2/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Parent 2
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/sub-taxonomy-1/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Sub taxonomy 1
                </span>
              </span>
            </a>
          </li>
          <li
            class="stacked-nav__list-item"
          >
            <a
              aria-current="false"
              href="/borderline-substances/parent-3/"
            >
              <span
                class="stacked-nav__content-wrapper"
              >
                <span>
                  Parent 3
                </span>
              </span>
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <div
      data-g="12 md:8 lg:9"
    >
      <div
        class="grid grid--rev grid--loose"
      >
        <div
          data-g="12 lg:3"
        >
          <nav
            aria-labelledby="on-this-page"
            class="wrapper"
          >
            <h2
              class="heading"
              id="on-this-page"
            >
              On this page
            </h2>
            <ol
              aria-label="Jump links to sections on this page"
              class="list"
            />
          </nav>
        </div>
        <div
          class="body"
          data-g="12 lg:9"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`BorderlineSubstancesSectionPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		bnfBorderlineSubstancesTaxonomy(id: { eq: $id }) {
			slug
			title
			rootTaxonomy {
				childTaxonomies {
					title
					childTaxonomies {
						title
						slug
					}
					slug
				}
			}
		}
	}
"
`;
